### PR TITLE
BUG: DatetimeIndex + arraylike of DateOffsets

### PR DESCRIPTION
**Pull request by jbrockmendel**
Before:

```
>>> dti = pd.date_range('2017-01-01', periods=2)
>>> other = np.array([pd.offsets.MonthEnd(), pd.offsets.Day(n=2)])

>>> dti + other
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc add cannot use operands with types dtype('<M8[ns]') and dtype('O')

# Same for `dti - other`, `dti + pd.Index(other)`, `dti - pd.Index(other)`

>>> dti + pd.Series(other)
0    DatetimeIndex(['2017-01-31', '2017-01-31'], dt...
1    DatetimeIndex(['2017-01-03', '2017-01-04'], dt...
dtype: object

# yikes.
```

After:
```
>>> dti + other
pandas/core/indexes/datetimelike.py:677: PerformanceWarning: Adding/subtracting array of DateOffsets to <class 'pandas.core.indexes.datetimes.DatetimeIndex'> not vectorized
  PerformanceWarning)
DatetimeIndex(['2017-01-31', '2017-01-04'], dtype='datetime64[ns]', freq=None)

>>> dti - pd.Index(other)
DatetimeIndex(['2016-12-31', '2016-12-31'], dtype='datetime64[ns]', freq=None)

>>> dti + pd.Series(other)
0   2017-01-31
1   2017-01-04
dtype: datetime64[ns]

```

<b>Caveat</b> This will need a follow-up to make sure `name` attribute is propogated correctly.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@jbrockmendel : As this PR stands, a little uneasy about it.  Besides the `name` bug that you mention as a "caveat," the fact that you have that `PerformanceWarning`, while not *super problematic*, is still a little bothersome.

File: pandas/tests/indexes/datetimes/test_arithmetic.py [hunk 1]
```diff
@@ -363,6 +363,46 @@ def test_datetimeindex_sub_timestamp_overflow(self):
             with pytest.raises(OverflowError):
                 dtimin - variant
 
+    @pytest.mark.parametrize('box', [np.array, pd.Index])
+    def test_dti_add_offset_array(self, tz, box):
+        dti = pd.date_range('2017-01-01', periods=2, tz=tz)
+        # TODO: check that `name` propogates correctly
```
Comment (gfyoung) [inline pandas/tests/indexes/datetimes/test_arithmetic.py]: Reference issue number below all of your new tests.

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -695,6 +706,23 @@ def __sub__(self, other):
...
             elif isinstance(other, Period):
                 return self._sub_period(other)
+            elif (isinstance(self, DatetimeIndex) and is_offsetlike(other) and
+                  not isinstance(other, ABCSeries)):
+                # Array of DateOffset objects
+                if len(other) == 1:
+                    return self - other[0]
+                else:
+                    from pandas.errors import PerformanceWarning
+                    warnings.warn("Adding/subtracting array of DateOffsets to "
```
Comment (gfyoung) [inline pandas/core/indexes/datetimelike.py]: Ah, I see that you added this yourself.  Still, feel a little uneasy about this.

**Comment by jbrockmendel** [comment]
> the fact that you have that PerformanceWarning, while not super problematic, is still a little bothersome.

This is explicitly copying the Series behavior.

**Comment by gfyoung** [comment]
> This is explicitly copying the Series behavior.

Hmmm...I see.  Theoretically, it makes sense.  Though if we can avoid adding this warning, that would be great.  Thus, either you can confirm that there is indeed a performance difference OR the code is rewritten so that we avoid this overhead.

**Comment by jbrockmendel** [comment]
> Thus, either you can confirm that there is indeed a performance difference OR the code is rewritten so that we avoid this overhead

I've been working on offsets a lot recently (and upcoming, see #18854).  Perf has improved quite a bit, but I can absolutely confirm that this is not a speedy operation.

File: pandas/core/indexes/datetimelike.py [hunk 3]
```diff
@@ -662,7 +663,21 @@ def __add__(self, other):
...
-            elif isinstance(other, (Index, datetime, np.datetime64)):
+            elif isinstance(other, (datetime, np.datetime64)):
+                return self._add_datelike(other)
+            elif (isinstance(self, DatetimeIndex) and is_offsetlike(other) and
+                  not isinstance(other, ABCSeries)):
+                # Array of DateOffset objects
+                if len(other) == 1:
+                    return self + other[0]
+                else:
+                    from pandas.errors import PerformanceWarning
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: so you are adding code here that already exists in ``datetimes.py:_add_offset`` ? is there a reason you are not dispatching to that (which may need to handle array-likes in addition to scalars). Further I am not in love with explicit type checking for DTI here. again why are you not handling that at a higher level.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimelike.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: _add_offset is for a single offset which does not have an implementation of `apply_index`.  This is for an array of offsets.  It's a case handled by Series but not by DatetimeIndex.

I can make a _add_offset_array and dispatch to that.

File: pandas/tests/indexes/datetimes/test_arithmetic.py [hunk 4]
```diff
@@ -363,6 +363,48 @@ def test_datetimeindex_sub_timestamp_overflow(self):
...
+    def test_dti_sub_offset_array(self, tz, box):
+        # GH#18824
+        dti = pd.date_range('2017-01-01', periods=2, tz=tz)
+        other = box([pd.offsets.MonthEnd(), pd.offsets.Day(n=2)])
+        res = dti - other
+        expected = DatetimeIndex([dti[n] - other[n] for n in range(len(dti))],
+                                 name=dti.name, freq='infer')
+        tm.assert_index_equal(res, expected)
+
+    def test_dti_with_offset_series(self, tz):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_arithmetic.py]: can you parametrize this with name (None, same name, other name) to make sure they are propagated correctly.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. just some additional testing requested around series + offset series names.

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -282,6 +282,7 @@ Conversion
 - Bug in :meth:`Index.astype` with a categorical dtype where the resultant index is not converted to a :class:`CategoricalIndex` for all types of index (:issue:`18630`)
 - Bug in :meth:`Series.astype` and ``Categorical.astype()`` where an existing categorical data does not get updated (:issue:`10696`, :issue:`18593`)
 - Bug in :class:`Series` constructor with an int or float list where specifying ``dtype=str``, ``dtype='str'`` or ``dtype='U'`` failed to convert the data elements to strings (:issue:`16605`)
+- Bug in :class:`DatetimeIndex` where adding or subtracting an array-like of ``DateOffset`` objects either raised (``np.array``, ``pd.Index``) or broadcast incorrectly (``pd.Series``) (:issue:`18224`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: use this PR number here (as 18224 is a very general reference). Is there any issue for this one specifically? (don't create one, just if there is an open one).

File: pandas/tests/indexes/datetimes/test_arithmetic.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_arithmetic.py]: Good call.  Looks like Series.op(Index) was always taking on the name of the Series.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**